---
title: "Differential expression alluvial plot"
author: "E. Roberson"
date: "April 27, 2018"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r libraries}
library( here )
library( ggalluvial )
library( tidyverse )
```

```{r shared_functions}
source( here::here( "project_shared_fxns.R" ) )
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(
	echo = TRUE,
	message = FALSE,
	fig.path = paste0( path_to_figures, "/" ),
	fig.keep = 'all',
	dpi = 300,
	fig.width = 11,
	fig.height = 8.5 )
```

# Section on Untreated vs. Control

```{r read_y_pull}
untreated_v_control <- read_csv( file = deg_file_untreated_control )

inactive_v_untreated <- read_csv( file = deg_file_inactive_untreated_paired )

untreated_v_control.gid <- direction_gene_ids_from_de( untreated_v_control )

inactive_v_untreated.gid <- inactive_v_untreated %>%
  mutate( FC = FC * ( -1.0 ) ) %>%
  direction_gene_ids_from_de( . )
# why, you might ask, are we flipping
# the fold-change?
# Reader, this was setup so that downstream
# we expect one part of alluvial to be
# case / control. And another part to be
# timepoint 2 / timepoint 1.
# i.e. untreated / control and
# inactive / untreated. Problem
# is the reviewers wanted it to be
# untreated / control and untreated / inactive
# so here we are. ¯\_(ツ)_/¯
```

DE Untreated v. Control: `r untreated_v_control.gid[[ 'num_de' ]]`.
DE Down: `r untreated_v_control.gid[[ 'num_down' ]]`.
DE Up: `r untreated_v_control.gid[[ 'num_up' ]]`.

DE Inactive v. Untreated: `r inactive_v_untreated.gid[[ 'num_de' ]]`.
DE Down: `r inactive_v_untreated.gid[[ 'num_down' ]]`.
DE Up: `r inactive_v_untreated.gid[[ 'num_up' ]]`.

```{r setup_alluvial_frame}
for_alluvial <- setup_alluvial_table_paired_data( untreated_v_control.gid, inactive_v_untreated.gid )

for_alluvial

for_alluvial <- for_alluvial %>%
  dplyr::rename( .data = ., count = n ) %>%
  filter( .data = ., count > 0 ) %>%
  filter( .data = ., !( Left == "Unchanged" & Right == "Unchanged" ) ) %>%
  as.data.frame( . )

left_sums <- for_alluvial %>%
  dplyr::group_by( .data = ., Left ) %>%
	dplyr::summarize(
		total = sum( count )
	) %>%
	as.data.frame( . )
rownames( left_sums ) = left_sums$Left

right_sums <- for_alluvial %>%
	dplyr::group_by( Right ) %>%
	dplyr::summarise(
		total = sum( count )
	) %>%
	as.data.frame( . )
rownames( right_sums ) = right_sums$Right

for_alluvial <- for_alluvial %>%
	mutate( Left_label = paste0( Left, "\n(n=", left_sums[ as.character( Left ), "total" ], ")" ) ) %>%
	mutate( Right_label = paste0( Right, "\n(n=", right_sums[ as.character( Right ), "total" ], ")" ) ) %>%
  write_csv( x = ., path = here::here( 'results', "alluvial_input_table_untreated.csv" ) )
```

```{r alluvial_plot}
ggplot( data = for_alluvial, aes( y = count, axis1 = Left_label, axis2 = Right_label ) ) +
  	geom_alluvium( aes( fill=Left ) ) +
  	geom_stratum( width = 1/4, fill = "black", color = "grey" ) +
  	geom_label( stat = "stratum", label.strata = TRUE ) +
  	theme_bw() +
  	gg_no_x_grid +
  	scale_x_continuous( breaks = 1:2, labels = c( 'Untreated', 'Inactive' ) ) +
  	gg_bigger_texts +
  	theme( legend.position = 'top' ) +
  	ggtitle( "Changes in gene expression in response to treatment" ) +
  	gg_center_title + 
    guides( fill = guide_legend( title = "" ) )
```

## Adding resolution status to the DEG output for Inactive vs. Untreated

```{r update_inactive_untreated_new_column}
# this logic was written with the original analysis
# i.e. the DEG were expressed as Inactive / Control
# now these lists were created above using that same logic
# remember we temporarily flipped the fold-change?
# so it's going to *seem* backwords, but is correct
# Is this a bunch of malarkey?
# yes it certainly is. but sometimes with reviews
# that suggest one "small" change the downstream 
# effects are many. this is one of them. again,
# not how I would have done it, but was necessary
# to address a single reviewer comment.
#
# TL;DR -- this is correct still, but seems backwards.
inactive_v_untreated %>%
	mutate( TreatmentEffect = case_when(
	  qval < 0.05 & FC < 0.0 & gene_id %in% untreated_v_control.gid[[ 'down_gene_ids' ]] ~ "Improved",
	  qval < 0.05 & FC > 0.0 & gene_id %in% untreated_v_control.gid[[ 'up_gene_ids' ]] ~ "Improved",
	  qval < 0.05 & FC < 0.0 & gene_id %in% untreated_v_control.gid[[ 'up_gene_ids' ]] ~ "Worsened",
	  qval < 0.05 & FC > 0.0 & gene_id %in% untreated_v_control.gid[[ 'down_gene_ids' ]] ~ "Worsened",
	  qval < 0.05 ~ "New",
	  qval >= 0.05 & gene_id %in% untreated_v_control.gid[[ 'up_gene_ids' ]] ~ "Unresolved",
	  qval >= 0.05 & gene_id %in% untreated_v_control.gid[[ 'down_gene_ids' ]] ~ "Unresolved",
	  qval >= 0.05 ~ "Never DE",
		TRUE ~ "Problem"
	)) %>%
	arrange( qval, pval, symbol ) %>%
	write_csv( x = ., path = str_replace( string = deg_file_inactive_untreated_paired, pattern = "_paired", replacement = "_paired_with_treateffect" ) )
```

# Section on JDM-only / Pre-psoriasis

```{r read_y_pull_PSOR}
prepsor_v_control <- read_csv( file = deg_file_prepsor_control )

prepsor.gid.list <- direction_gene_ids_from_de( prepsor_v_control )

psor_v_prepsor <- read_csv( file = deg_file_prepsor_psor_paired )

psor.gid.list <- direction_gene_ids_from_de( psor_v_prepsor )
```

DE PrePsor: `r prepsor.gid.list[[ 'num_de' ]]`.
DE Down: `r prepsor.gid.list[[ 'num_down' ]]`.
DE Up: `r prepsor.gid.list[[ 'num_up' ]]`.

DE Psor vs. PrePsor: `r psor.gid.list[[ 'de' ]]`. 
DE Down: `r psor.gid.list[[ 'num_down' ]]`. 
DE Up: `r psor.gid.list[[ 'num_up' ]]`.

```{r setup_alluvial_frame_PSOR}
for_alluvial <- setup_alluvial_table_paired_data( prepsor.gid.list, psor.gid.list )

as.data.frame( for_alluvial )

for_alluvial <- for_alluvial %>%
	filter( n > 0 ) %>%
	filter( !( Left == "Unchanged" & Right == "Unchanged") )

left_sums <- for_alluvial %>%
	dplyr::group_by( Left ) %>%
	dplyr::summarise(
		n = sum( n )
	) %>%
	as.data.frame( . )
rownames( left_sums ) = left_sums$Left

right_sums <- for_alluvial %>%
	dplyr::group_by( Right ) %>%
	dplyr::summarise(
		n = sum( n )
	) %>%
	as.data.frame( . )
rownames( right_sums ) = right_sums$Right

for_alluvial <- for_alluvial %>%
	mutate( Left_label = paste0( Left, "\n(n=", left_sums[ as.character( Left ), "n" ], ")" ) ) %>%
	mutate( Right_label = paste0( Right, "\n(n=", right_sums[ as.character( Right ), "n" ], ")" ) ) %>%
  write_csv( x = ., path = here::here( 'results', "alluvial_input_table_psoriasis.csv" ) )
```

```{r alluvial_plot_PSOR}
ggplot( data = for_alluvial, aes( y = n, axis1 = Left_label, axis2 = Right_label ) ) +
	geom_alluvium( aes( fill=Left ) ) +
	geom_stratum( width = 1/4, fill = "black", color = "grey" ) +
	geom_label( stat = "stratum", label.strata = TRUE ) +
	theme_bw() +
	gg_no_x_grid +
	scale_x_continuous( breaks = 1:2, labels = c( 'PrePsoriasis', 'Psoriasis' ) ) +
	gg_bigger_texts +
	theme( legend.position = 'top' ) +
	ggtitle( "JDM-only vs. Psoriasis" ) +
	gg_center_title + 
    guides( fill = guide_legend( title = "JDM-only" ) )
```

# Finish up with version info

```{r session_information}
Sys.time()

getwd()
  
sessionInfo()
```
